### PR TITLE
Feature/initial integration

### PR DESCRIPTION
**Pull request by eXtremeX**
## Summary

- Ingest new entities
  - `jenkins_account`
  - `jenkins_build`
  - `jenkins_job`
  - `jenkins_repository`
  - `jenkins_role`
  - `jenkins_user`
  
- Build new relationships
  - `jenkins_account_has_job`
  - `jenkins_account_has_role`
  - `jenkins_account_has_user`
  - `jenkins_build_has_repository`
  - `jenkins_job_has_build`
  - `jenkins_job_has_job`
  - `jenkins_user_assigned_role`
  - `jenkins_user_has_build`

## Type of change

Please leave any irrelevant options unchecked.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
      not work as expected)
- [x] This change requires a documentation update

### General Development Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### Integration Development Checklist:

Please leave any irrelevant options unchecked.

- [x] I have checked for additional permissions required to call any new API
      endpoints, and have documented any additional permissions in
      `jupiterone.md`, where necessary.
- [x] My changes properly paginate the target service provider's API
- [x] My changes properly handle rate limiting of the target service provider's
      API
- [x] My new integration step is instrumented to execute in the correct order
      using `dependsOn`
- [x] I have referred to the
      [JupiterOne data model](https://github.com/JupiterOne/data-model/tree/main/src/schemas)
      to ensure that any new entities/relationships, and relevant properties,
      match the recommended model for this class of data
- [x] I have updated the `CHANGELOG.md` file to describe my changes
- [x] When changes include modifications to existing graph data ingestion, I've
      reviewed all existing managed questions referencing the entities,
      relationships, and their property names, to ensure those questions still
      function with my changes.

### CODE REVIEW

File: src/steps/build/converter.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+export function createBuildEntity(build: JenkinsBuild): Entity {
+  const startedByUserId = build.actions[0].causes[0].userId;
+  let remoteUrl;
+  for (const index in build.actions || []) {
+    if (build.actions[index]._class == 'hudson.plugins.git.util.BuildData') {
+      const remoteUrlTemp = build.actions[index].remoteUrls;
+      remoteUrl = remoteUrlTemp;
+      break;
+    }
+  }
```
Comment (neurovelho) [inline src/steps/build/converter.ts:22]: Could be replaced with `arr.find`.

**Review by neurovelho — APPROVED** [review]

File: src/steps/user/index.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+export async function fetchUsers({
+  instance,
+  jobState,
+}: IntegrationStepExecutionContext<IntegrationConfig>) {
+  const apiClient = createAPIClient(instance.config);
+  const accountEntity = (await jobState.getData(ACCOUNT_ENTITY_KEY)) as Entity;
+
+  await apiClient.iterateUsers(async (user) => {
+    const userEntity = await jobState.addEntity(createUserEntity(user));
```
Comment (austinkelleher) [inline src/steps/user/index.ts:29]: This is producing a duplicate key against our prod Jenkins instance. Is there a more unique `_key` than this? https://github.com/JupiterOne/graph-jenkins/pull/4/files#diff-af8ebd7417e323249847fbfa1033c3d21088dafe73fc1d4b3c087454bed8570cR15

**Review by austinkelleher — CHANGES_REQUESTED** [review]

File: src/client.ts [hunk 3]
```diff
@@ -1,121 +1,236 @@
...
+    return await this.getRequest(`${url}/api/json`, 'GET');
+  }
+
+  public async listAllRoles(): Promise<any> {
+    const roleList: Array<string> = [];
+
+    for (const roleType of this.roleTypes) {
+      const roles = await this.getRequest(
+        this.withBaseUri(
+          `/role-strategy/strategy/getAllRoles?type=${roleType}`,
```
Comment (austinkelleher) [inline src/client.ts:198]: I'm seeing a 403 when running this against our prod Jenkins instance with my user. I wonder if this API requires admin permissions? If so, can we please document that and ensure that this step emits a job log event that indicates the necessary permission?

File: src/steps/repository/converter.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+export function createRepositoryEntity(repository: JenkinsRepository): Entity {
+  return createIntegrationEntity({
+    entityData: {
+      source: repository,
+      assign: {
+        _type: Entities.REPOSITORY._type,
+        _class: Entities.REPOSITORY._class,
+        _key: getRepositoryKey(repository.url),
+        name: repository.url,
+        url: repository.url,
```
Comment (austinkelleher) [inline src/steps/repository/converter.ts:22]: Would be nice if we could build a `webLink` from this. I understand that these can be Git URLs.

File: src/steps/job/converter.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  return createIntegrationEntity({
+    entityData: {
+      source: project,
+      assign: {
+        _type: Entities.JOB._type,
+        _class: Entities.JOB._class,
+        _key: getJobKey(`${project.url}-${project.name}`),
+        id: project.name,
+        name: project.name,
+        url: project.url,
```
Comment (austinkelleher) [inline src/steps/job/converter.ts:23]:
```suggestion
        webLink: project.url,
```

File: .github/workflows/gitleaks.yml [hunk 6]
```diff
@@ -11,4 +11,4 @@ jobs:
     steps:
       - uses: actions/checkout@v1
       - name: gitleaks-action
-        uses: zricethezav/gitleaks-action@master
+        uses: gitleaks/gitleaks-action@v1.6.0
```
Comment (austinkelleher) [inline .github/workflows/gitleaks.yml:14]: FYI we now are using GitHub Enterprise which has built-in secrets scanning. We can completely remove this file and the top level `.gitleaks.yml` file

Here is an example PR: https://github.com/JupiterOne/graph-asana/pull/11

File: docs/development.md [hunk 7]
```diff
@@ -1,28 +1,24 @@
...
 
 ## Provider account setup
 
-Please provide information about the steps needed to create an account with a
-provider. Images and references to a provider's documentation is very helpful
-for new developers picking up your work.
+1. Install Jenkins on the local machine/server and take note of the provided
+   domain
+2. Install recommended plugins
+3. Create an admin account on this installation
```
Comment (austinkelleher) [inline docs/development.md:12]: I don't know whether an admin is truly necessary. My user in our prod Jenkins instance is _not_ an admin and was able to collect most of the data. I made a comment below about the inability to collect role data.

File: src/steps/user/converter.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+  return createIntegrationEntity({
+    entityData: {
+      source: users,
+      assign: {
+        _type: Entities.USER._type,
+        _class: Entities.USER._class,
+        _key: getUserKey(userId),
+        id: userId,
+        lastChange: users.lastChange,
```
[the remark this answers is not in the document]
Comment (eXtremeX) [reply src/steps/user/converter.ts:25]: Looks that way, however I wasn't able to see any value in the lastChange value after modifying my user :thinking: 
Do you happen to have a raw data that shows what that value might look like? (I'd want to use the helper `parseTimePropertyValue` function if possible)

[the remark this answers is not in the document]
Comment (eXtremeX) [reply src/steps/user/converter.ts:25]: Update: saw response blob over Slack, so no longer needed

**Comment by neurovelho** [comment]
@austinkelleher Just FYI, I added a Dockerfile (modified from graph-kubernetes), that can be used to build an image capable of running this integration in a Jenkins pipeline.

Here's an example pipeline:

```
pipeline {
    agent {
        docker { image 'creativicedev/graph-jenkins:0.0.0' }
    }
    
    environment {
        JUPITERONE_API_KEY = 'j1_api_key'
        JUPITERONE_ACCOUNT = 'j1_account'
        USER_NAME = 'jenkins_username'
        API_KEY   = 'jenkins_api_key'
        HOST_NAME = 'http://172.17.0.1:8080'
    }
    
    stages {
        stage('Test') {
            steps {
                sh 'cd /opt/jupiterone/integration && yarn start'
            }
        }
    }
}
```

File: docs/jupiterone.md [hunk 9]
```diff
@@ -39,42 +30,64 @@ If you need help with this integration, please contact
 
 ## Integration Walkthrough
 
-### In {{provider}}
-
-TODO: List specific actions that must be taken in the provider. Remove this
-section when there are no actions to take in the provider.
-
-1. [Generate a REST API key](https://example.com/docs/generating-api-keys)
+### In Jenkins
```
Comment (austinkelleher) [inline docs/jupiterone.md:33]: 👍 Thanks for adding docs about this!

**Review by austinkelleher — APPROVED** [review]

File: docs/spec/src/build/index.ts [hunk 10]
```diff
@@ -1,49 +1,49 @@
...
-        _type: 'jenkins_user_has_build',
-        sourceType: 'jenkins_user',
-        _class: RelationshipClass.HAS,
-        targetType: 'jenkins_build',
-      },
-    ],
-    dependsOn: ['fetch-build', 'fetch-users'],
-    implemented: true,
-  },
+  // {
```
Comment (VDubber) [inline docs/spec/src/build/index.ts:5]: @eXtremeX From a team meeting, I believe we will never want to support the ingestion of jenkin jobs - too much data :) 

Please remove all commented code relating to job ingestion

**Comment by mdaum** [comment]
Thanks for all this work! Looks like we need to merge `main` into this branch and we should be good to go!. I just finished adding image build/publish workflow setup for this repo. So an image should be published when we merge this to `main`, and when we cut a new version 👍

**Review by adam-in-ict — APPROVED** [review]

**Pull request merged**